File: web/src/lib/api/scheduling.ts
import { request } from "./client";
import type {
  NeedsAttentionItem,
  ScheduleApplyItemInput,
  ScheduleApplyResult,
  ScheduleProposal,
  ScheduleRequestInput,
} from "./types";

/**
 * Proposes a schedule -- never writes anything to Google Calendar and
 * never persists the proposal. "Plan my day"/"Plan my work" calls this
 * with no task_ids to consider every unscheduled, already-prioritized task.
 */
export function proposeSchedule(input: ScheduleRequestInput = {}): Promise<ScheduleProposal> {
  return request<ScheduleProposal>("/tasks/schedule", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

/**
 * Applies approved schedule items to Google Calendar. The backend
 * revalidates every item against fresh availability and the task's real
 * deadline before writing anything -- these start/end values are a
 * request, not a fact the backend trusts blindly.
 */
export function applySchedule(items: ScheduleApplyItemInput[]): Promise<ScheduleApplyResult> {
  return request<ScheduleApplyResult>("/tasks/schedule/apply", {
    method: "POST",
    body: JSON.stringify({ items }),
  });
}

/**
 * Previously-applied schedule items whose Google Calendar event was
 * deleted externally (Phase 7 two-way sync) -- never auto-recreated;
 * re-applying the task creates a fresh event and clears this.
 */
export function listNeedsAttention(): Promise<NeedsAttentionItem[]> {
  return request<NeedsAttentionItem[]>("/tasks/schedule/needs-attention");
}

File: backend/app/schemas/scheduling.py
"""Scheduling API request/response shapes. See app/services/scheduling.py
for the actual algorithm -- these models are just its I/O boundary."""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel


class ScheduleRequest(BaseModel):
    """Omit `task_ids` to consider every unscheduled, already-prioritized
    task in the tenant (status='pending' with at least one AI result).
    Omit `horizon_start`/`horizon_end` to default to now .. now+14 days."""

    task_ids: list[uuid.UUID] | None = None
    horizon_start: datetime | None = None
    horizon_end: datetime | None = None


class ProposedScheduleItem(BaseModel):
    task_id: uuid.UUID
    title: str
    start: datetime
    end: datetime
    priority_score: float
    score: float
    reason: str


class UnscheduledTaskOut(BaseModel):
    task_id: uuid.UUID
    title: str
    reason: str


class ScheduleProposal(BaseModel):
    horizon_start: datetime
    horizon_end: datetime
    scheduled: list[ProposedScheduleItem]
    unscheduled: list[UnscheduledTaskOut]


class ScheduleApplyItem(BaseModel):
    """One item the user approved from a proposal. The backend re-validates
    `start`/`end` against the task's real deadline/duration and fresh
    calendar availability before ever writing anything -- these values are
    a *request*, not a fact (see /docs/architecture.md "Apply endpoint
    revalidation"; never trusted blindly per the phase brief)."""

    task_id: uuid.UUID
    start: datetime
    end: datetime


class ScheduleApplyRequest(BaseModel):
    items: list[ScheduleApplyItem]


class AppliedItemStatus(str, Enum):
    created = "created"
    already_applied = "already_applied"
    failed = "failed"


class AppliedItemResult(BaseModel):
    task_id: uuid.UUID
    status: AppliedItemStatus
    google_event_id: str | None = None
    start: datetime | None = None
    end: datetime | None = None
    reason: str | None = None


class ScheduleApplyResult(BaseModel):
    created: int
    already_applied: int
    failed: int
    results: list[AppliedItemResult]


class ScheduleItemOut(BaseModel):
    """One applied schedule item in a date range -- the source of truth
    for "what's on my plan today" (mobile Today/Calendar screens), joining
    in just enough from the task/AI-result/Calendar-mapping tables that
    clients don't have to make three separate calls and stitch it
    together themselves."""

    schedule_item_id: uuid.UUID
    task_id: uuid.UUID
    title: str
    starts_at: datetime
    ends_at: datetime
    status: str
    needs_attention: bool
    attention_reason: str | None = None
    google_event_id: str | None = None
    sync_status: str | None = None
    priority_score: float | None = None


class NeedsAttentionItemOut(BaseModel):
    """A previously-applied schedule item whose Google Calendar event was
    deleted externally (see app/services/calendar_sync.py). The app never
    silently recreates it -- this is how it's surfaced instead; re-applying
    the task (POST /tasks/schedule/apply) creates a fresh event and clears
    this."""

    task_id: uuid.UUID
    schedule_item_id: uuid.UUID
    title: str
    reason: str | None = None
    starts_at: datetime
    ends_at: datetime
